src/Razor/src/Microsoft.VisualStudio.LanguageServerClient.Razor/LSPRazorProjectHost.cs

```diff
@@ -0,0 +1,124 @@
+            [ImportMany] IEnumerable<Lazy<ILanguageClient, IDictionary<string, object>>> languageClients)
+        {
+            if (joinableTaskContext is null)
+            {
+                throw new ArgumentNullException(nameof(joinableTaskContext));
+            }
+
+            if (featureDetector is null)
+            {
+                throw new ArgumentNullException(nameof(featureDetector));
+            }
+
+            if (languageClientBroker is null)
+            {
+                throw new ArgumentNullException(nameof(languageClientBroker));
+            }
+
+            if (languageClients is null)
+            {
+                throw new ArgumentNullException(nameof(languageClients));
+            }
+
+            _joinableTaskFactory = joinableTaskContext.Factory;
+            _featureDetector = featureDetector;
+            _languageClientBroker = languageClientBroker;
+            _applicableLanguageClients = GetApplicableClients(languageClients);
+        }
+
+        public async Task LoadAsync()
+        {
+            await _joinableTaskFactory.SwitchToMainThreadAsync();
+
+            if (!_featureDetector.IsLSPEditorFeatureEnabled())
+            {
+                return;
+            }
+
+            foreach (var client in _applicableLanguageClients)
+            {
+                _languageClientBroker.Value.LoadAsync(client.Metadata, client.Value).Forget();
+            }
+        }
+
+        public Task UnloadAsync()
+        {
+            return Task.CompletedTask;
+        }
+
+        private static List<Lazy<ILanguageClient, ILanguageClientMetadata>> GetApplicableClients(IEnumerable<Lazy<ILanguageClient, IDictionary<string, object>>> languageClients)
+        {
+            var applicableContentTypes = new[]
+            {
+                RazorLSPConstants.RazorLSPContentTypeName,
+                RazorLSPConstants.CSharpLSPContentTypeName,
+                RazorLSPConstants.HtmlLSPContentTypeName,
+            };
+
+            var applicableClients = new List<Lazy<ILanguageClient, ILanguageClientMetadata>>();
+            foreach (var client in languageClients)
+            {
+                if (!client.Metadata.TryGetValue(nameof(ILanguageClientMetadata.ContentTypes), out var contentTypeValue) ||
+                    !(contentTypeValue is IEnumerable<string> contentTypes) ||
+                    !contentTypes.Intersect(applicableContentTypes).Any())
+                {
+                    continue;
+                }
+
+                string clientName = null;
+                if (client.Metadata.TryGetValue(nameof(ILanguageClientMetadata.ClientName), out var clientNameValue))
+                {
+                    clientName = clientNameValue.ToString();
+                }
+
+                applicableClients.Add(new Lazy<ILanguageClient, ILanguageClientMetadata>(
+                    () => { return client.Value; },
+                    new LanguageClientMetadata(clientName, contentTypes)));
```

Comment:
What does it mean to have a `null` clientName?

Reply:
Not sure if we need this check, LSP platform doesn't have it. @ajaybhargavb was there another reason for adding it / should there be additional handling in case of `null`?

Reply:
I was trying to follow the same pattern followed by the LSP platform so we don't introduce any new surprises. Having said that, you might be able to simplify this entire method if we can just use the `ILanguageClientMetadata` type directly. See this [comment](https://github.com/dotnet/aspnetcore-tooling/pull/2164#discussion_r450430977).

Reply:
N/A leaving as is. https://github.com/dotnet/aspnetcore-tooling/pull/2164#discussion_r450529926